src/libexpr/primops.cc

```diff
@@ -4074,54 +4093,210 @@ void EvalState::createBaseEnv()
-    addConstant("null", v);
+    addConstant("null", v, {
+        .type = nNull,
+        .doc = R"(
+          Primitive value.
+
+          The name `null` is not special, and can be shadowed:
+
+          ```nix-repl
+          nix-repl> let null = 1; in null
+          1
+          ```
+        )",
+    });
 
     if (!evalSettings.pureEval) {
         v.mkInt(time(0));
-        addConstant("__currentTime", v);
+    }
+    addConstant("__currentTime", v, {
+        .type = nInt,
+        .doc = R"(
+          Return the [Unix time](https://en.wikipedia.org/wiki/Unix_time) at first evaluation.
+          Repeated references to that name will re-use the initially obtained value.
+
+          Example:
 
+          ```console
+          $ nix repl
+          Welcome to Nix 2.15.1 Type :? for help.
+
+          nix-repl> builtins.currentTime
+          1683705525
+
+          nix-repl> builtins.currentTime
+          1683705525
+          ```
+
+          The [store path](@docroot@/glossary.md#gloss-store-path) of a derivation depending on `currentTime` will differ for each evaluation.
+        )",
+        .impureOnly = true,
+    });
+
+    if (!evalSettings.pureEval) {
         v.mkString(settings.thisSystem.get());
-        addConstant("__currentSystem", v);
     }
+    addConstant("__currentSystem", v, {
+        .type = nString,
+        .doc = R"(
+          A string in the form `<cpu-architecture>-<operating-systems>` identifying the "current" system.
+          The current system is by default the system where the Nix expression is being evaluated.
+
+          Example of possible `builtins.currentSystem` values:
+
+          - `"i686-linux"`
+          - `"x86_64-darwin"`.
+
+          This builtin constant exists so a Nix expression can be implicitly parameterized what system for which derivations produced by evaluation should be built.
+          If all derivations produced by expression have their system field set with ``builtins.currentSystem` like this:
+
+          ```nix
+          builtins.derivationStrict {
+             ...
+             system = builtins.currentSystem;
+          }
+          ```
+
+          Then, by default, those derivations can be built on the same platform where the expression that produced them was evaluated.
+          This pattern makes this case of evaluating and building on the same machine more ergonomic than e.g. having to pass an argument for the current system on the command-line.
+
+          Pure evaluation eschews this because implicit parameterization like this is impure.
+          The problem of making evaluating and building on the current machine ergonomic can be solved in other ways.
+
+          When available, the current system is defined as the value of the [system](@docroot@/command-ref/conf-file.md#conf-pure-eval) configuration option.
+          That configuration option can be overridden in order to evaluate derivations for different system than the current one instead:
```

Comment:
Why is this so much text?  All of the information should be in one place, where the value itself is defined. https://github.com/NixOS/nix/pull/8315/files Please add suggestions there.


```suggestion
          The value of the [`system` configuration option](@docroot@/command-ref/conf-file.md#conf-pure-eval).
          
          It can be used to set the `system` attribute for `builtins.derivation` such that the resulting derivation can be built on the same system that evaluates the Nix expression:
          
          ```nix
           builtins.derivation {
             # ...
             system = builtins.currentSystem;
          }
          ```

          It can be overridden in order to create derivations for different system than the current one:
```